Code/GraphMol/RGroupDecomposition/RGroupDecompParams.cpp

```diff
@@ -310,50 +315,66 @@ void RGroupDecompositionParameters::addDummyAtomsToUnlabelledCoreAtoms(
     return;
   }
 
-  // add a single rgroup substitution to any atom with free valence that doesn't
-  // have a terminal dummy neighbor
-
-  std::vector<Atom *> unlabelledCoreAtoms{};
+  // add R group substitutions to fill atomic valence
+  std::vector<Atom *> unlabeledCoreAtoms{};
   for (const auto atom : core.atoms()) {
     if (atom->getAtomicNum() == 1) {
       continue;
     }
-
     if (hasLabel(atom, labels)) {
       continue;
     }
+    unlabeledCoreAtoms.push_back(atom);
+  }
 
-    if (atom->getAtomicNum() > 0 && atom->calcImplicitValence() <= 1 &&
-        !atom->hasQuery()) {
-      continue;
-    }
+  for (const auto atom : unlabeledCoreAtoms) {
+    atom->calcImplicitValence(false);
+    const auto atomIndex = atom->getIdx();
+    int dummiesToAdd;
+    int maxNumDummies = 4 - static_cast<int>(atom->getDegree());
 
-    auto [nbrIdx, endNbrs] = core.getAtomNeighbors(atom);
-    bool hasTerminalDummyNeighbor = false;
+    // figure out the number of dummies to add
+    if (atom->getAtomicNum() == 0) {
+      dummiesToAdd = maxNumDummies;
+    } else {
+      double bondOrder = 0;
```

Comment:
I'm not sure I follow the logic of the below code.
Can't you just use the total number of implicit Hs here?

Reply:
Not if there is a query bond on the atom and the default smarts bond is single or aromatic.  In this core (from *testGibhub5613*) `C(=O)*C1~C~C~*~*~C~1` there are no implicit hydrogens on the atoms with query bonds.   I guess we could use number of implicit hydrogens if none of the neighbor bonds were queries.

Reply:
ah, right, that makes sense.